File: frameworks/quarkus-graphql/src/main/java/com/velocitybench/repository/UserRepository.java
package com.velocitybench.repository;

import com.velocitybench.model.User;
import io.agroal.api.AgroalDataSource;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;

import java.sql.*;
import java.util.*;

@ApplicationScoped
public class UserRepository {

    @Inject
    AgroalDataSource dataSource;

    public Optional<User> findById(String id) {
        String sql = "SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user WHERE id = ?";
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setObject(1, UUID.fromString(id));
            try (ResultSet rs = stmt.executeQuery()) {
                if (rs.next()) {
                    return Optional.of(mapRow(rs));
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return Optional.empty();
    }

    public Map<Integer, User> findByPks(Set<Integer> pks) {
        if (pks.isEmpty()) return Collections.emptyMap();

        String placeholders = String.join(",", Collections.nCopies(pks.size(), "?"));
        String sql = "SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user WHERE pk_user IN (" + placeholders + ")";

        Map<Integer, User> result = new HashMap<>();
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            int i = 1;
            for (Integer pk : pks) {
                stmt.setInt(i++, pk);
            }
            try (ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    User user = mapRow(rs);
                    result.put(user.getPkUser(), user);
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return result;
    }

    public List<User> findAll(int limit) {
        String sql = "SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user ORDER BY pk_user LIMIT ?";
        List<User> users = new ArrayList<>();
        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            stmt.setInt(1, Math.min(limit, 100));
            try (ResultSet rs = stmt.executeQuery()) {
                while (rs.next()) {
                    users.add(mapRow(rs));
                }
            }
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }
        return users;
    }

    public Optional<User> update(String id, String fullName, String bio) {
        StringBuilder sql = new StringBuilder("UPDATE tb_user SET updated_at = NOW()");
        List<Object> params = new ArrayList<>();

        if (fullName != null) {
            sql.append(", full_name = ?");
            params.add(fullName);
        }
        if (bio != null) {
            sql.append(", bio = ?");
            params.add(bio);
        }

        sql.append(" WHERE id = ?");
        params.add(UUID.fromString(id));

        try (Connection conn = dataSource.getConnection();
             PreparedStatement stmt = conn.prepareStatement(sql.toString())) {
            for (int i = 0; i < params.size(); i++) {
                stmt.setObject(i + 1, params.get(i));
            }
            stmt.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException(e);
        }

        return findById(id);
    }

    private User mapRow(ResultSet rs) throws SQLException {
        return new User(
            rs.getInt("pk_user"),
            rs.getString("id"),
            rs.getString("username"),
            rs.getString("full_name"),
            rs.getString("bio"),
            rs.getTimestamp("created_at").toLocalDateTime(),
            rs.getTimestamp("updated_at") != null ? rs.getTimestamp("updated_at").toLocalDateTime() : null
        );
    }
}

File: frameworks/play-graphql/app/repositories/UserRepository.scala
package repositories

import db.Database
import models.User
import javax.inject.{Inject, Singleton}
import java.util.UUID
import scala.collection.mutable

@Singleton
class UserRepository @Inject()(database: Database) {

  def findById(id: String): Option[User] = {
    val conn = database.getConnection
    try {
      val stmt = conn.prepareStatement(
        "SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user WHERE id = ?"
      )
      stmt.setObject(1, UUID.fromString(id))
      val rs = stmt.executeQuery()
      if (rs.next()) Some(mapRow(rs)) else None
    } finally {
      conn.close()
    }
  }

  def findByPks(pks: Set[Int]): Map[Int, User] = {
    if (pks.isEmpty) return Map.empty

    val conn = database.getConnection
    try {
      val pkSeq = pks.toSeq
      val placeholders = Seq.fill(pkSeq.size)("?").mkString(",")
      val stmt = conn.prepareStatement(
        s"SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user WHERE pk_user IN ($placeholders)"
      )
      pkSeq.zipWithIndex.foreach { case (pk, idx) => stmt.setInt(idx + 1, pk) }
      val rs = stmt.executeQuery()
      val result = mutable.Map[Int, User]()
      while (rs.next()) {
        val user = mapRow(rs)
        result(user.pkUser) = user
      }
      result.toMap
    } finally {
      conn.close()
    }
  }

  def findAll(limit: Int): Seq[User] = {
    val conn = database.getConnection
    try {
      val stmt = conn.prepareStatement(
        "SELECT pk_user, id, username, full_name, bio, created_at, updated_at FROM tb_user ORDER BY pk_user LIMIT ?"
      )
      stmt.setInt(1, Math.min(limit, 100))
      val rs = stmt.executeQuery()
      val users = mutable.Buffer[User]()
      while (rs.next()) {
        users += mapRow(rs)
      }
      users.toSeq
    } finally {
      conn.close()
    }
  }

  def update(id: String, fullName: Option[String], bio: Option[String]): Option[User] = {
    val conn = database.getConnection
    try {
      val sets = mutable.Buffer("updated_at = NOW()")
      val params = mutable.Buffer[Any]()

      fullName.foreach { fn =>
        sets += "full_name = ?"
        params += fn
      }
      bio.foreach { b =>
        sets += "bio = ?"
        params += b
      }

      val sql = s"UPDATE tb_user SET ${sets.mkString(", ")} WHERE id = ?"
      val stmt = conn.prepareStatement(sql)
      params.zipWithIndex.foreach { case (p, idx) => stmt.setObject(idx + 1, p) }
      stmt.setObject(params.size + 1, UUID.fromString(id))
      stmt.executeUpdate()
    } finally {
      conn.close()
    }
    findById(id)
  }

  private def mapRow(rs: java.sql.ResultSet): User = {
    User(
      pkUser = rs.getInt("pk_user"),
      id = java.util.UUID.fromString(rs.getString("id")),
      username = rs.getString("username"),
      fullName = Option(rs.getString("full_name")),
      bio = Option(rs.getString("bio")),
      createdAt = rs.getTimestamp("created_at").toLocalDateTime,
      updatedAt = Option(rs.getTimestamp("updated_at")).map(_.toLocalDateTime)
    )
  }
}
